onnxruntime/core/xnnpack/optimizer/xnnpack_transformer.cc

```diff
@@ -0,0 +1,497 @@
+using namespace ::onnxruntime::common;
+using namespace onnx_layout_transformation;
+
+#define XNNPACK_RETURN_IF_ERROR(expr)                                                                                 \
+  do {                                                                                                                \
+    auto _status = (expr);                                                                                            \
+    if ((!_status.IsOK())) {                                                                                          \
+      std::ostringstream oss;                                                                                         \
+      oss << __FILE__ << ":" << __LINE__ << " Convert the model to XNNPack graph failed: " << _status.ErrorMessage(); \
+      return Status(_status.Category(), _status.Code(), oss.str());                                                   \
+    }                                                                                                                 \
+  } while (0)
+
+namespace onnxruntime {
+static bool IsPaddingTypeSupported(AutoPadType auto_pad) {
+  return auto_pad == AutoPadType::NOTSET || auto_pad == AutoPadType::VALID || auto_pad == AutoPadType::SAME_UPPER;
+}
+
+bool IsMaxPoolSupportedByXNNPack(const Node& nodeRef, bool input_is_nchw) {
+  if (nodeRef.OpType() != "MaxPool") return false;
+  auto input_defs = nodeRef.InputDefs();
+  if (input_defs.size() != 1) return false;
+  if (!input_defs[0]->HasTensorOrScalarShape()) return false;
+  auto X_input = input_defs[0]->TypeAsProto();
+  assert(X_input != nullptr);
+  int32_t etype = X_input->tensor_type().elem_type();
+  // Currently we only implemented a f32 kernel
+  // if (etype != ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT && etype != ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8 && etype !=
+  // ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT8) return false;
+  if (etype != ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT) return false;
+  ProtoHelperNodeContext nc(nodeRef);
+  OpNodeProtoHelper info(&nc);
+  PoolAttributes pool_attrs(info, "MaxPool", nodeRef.SinceVersion());
+  if (!IsPaddingTypeSupported(pool_attrs.auto_pad)) return false;
+  if (pool_attrs.kernel_shape.size() != 2) return false;
+  if (pool_attrs.kernel_shape[0] <= 0 || pool_attrs.kernel_shape[1] <= 0) return false;
+  if (pool_attrs.kernel_shape[0] == 1 && pool_attrs.kernel_shape[1] == 1) {
+    // XNNPack doesn't like to support 1x1 maxpool.
+    return false;
+  }
+  auto& input_shape = X_input->tensor_type().shape();
+  if (input_shape.dim_size() != 4) return false;
+  auto& channel_dim = input_is_nchw ? input_shape.dim(1) : input_shape.dim(3);
+  if (!channel_dim.has_dim_value()) return false;
+  return true;
+}
+
+Status IsConvSupportedByXNNPack(const Node& nodeRef, std::unordered_set<const NodeArg*>& graph_const_values,
+                                bool input_is_nchw) {
+  if (nodeRef.OpType() != "Conv") return Status(ONNXRUNTIME, FAIL);
```

Comment:
>Conv [](http://example.com/codeflow?start=27&length=4)

we need to check the domain to be ai.onnx too, for this:
[https://github.com/microsoft/onnxruntime/blob/47c09e670126b33a57491627006d7d447754373e/onnxruntime/core/graph/contrib_ops/nchwc_schema_defs.cc#L149](https://github.com/microsoft/onnxruntime/blob/47c09e670126b33a57491627006d7d447754373e/onnxruntime/core/graph/contrib_ops/nchwc_schema_defs.cc#L149)

Similar to MaxPool